pkg/gardenlet/controller/federatedseed/seed_control.go

```diff
@@ -0,0 +1,252 @@
+)
+
+// Controller is responsible for maintaining multiple federated Seeds' controllers
+type Controller struct {
+	federatedSeedControllerManager *federatedSeedControllerManager
+	k8sGardenClient                kubernetes.Interface
+
+	config *config.GardenletConfiguration
+
+	seedLister    gardencorelisters.SeedLister
+	seedHasSynced cache.InformerSynced
+	seedQueue     workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+
+	numberOfRunningWorkers int
+	workerCh               chan int
+}
+
+// NewFederatedSeedController creates new controller that syncs extensions states to the ShootState resource in the garden cluster.
+func NewFederatedSeedController(k8sGardenClient kubernetes.Interface, gardenCoreInformerFactory gardencoreinformers.SharedInformerFactory, config *config.GardenletConfiguration, recorder record.EventRecorder) *Controller {
+	var seedInformer = gardenCoreInformerFactory.Core().V1beta1().Seeds()
+
+	controller := &Controller{
+		k8sGardenClient: k8sGardenClient,
+		seedLister:      seedInformer.Lister(),
+		seedQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "seed-controllers"),
+		config:          config,
+		workerCh:        make(chan int),
+		recorder:        recorder,
+	}
+
+	controller.federatedSeedControllerManager = &federatedSeedControllerManager{
+		seedShootStateSyncController: make(map[string]*shootstate.SyncController),
+	}
+
+	seedInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controllerutils.SeedFilterFunc(confighelper.SeedNameFromSeedConfig(config.SeedConfig), config.SeedSelector),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.seedAdd,
+			UpdateFunc: controller.seedUpdate,
+			DeleteFunc: controller.seedDelete,
+		},
+	})
+
+	controller.seedHasSynced = seedInformer.Informer().HasSynced
+	return controller
+}
+
+func (c *Controller) seedAdd(obj interface{}) {
+	seedObj := obj.(*gardencorev1beta1.Seed)
+	bootstrappedCondition := helper.GetCondition(seedObj.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+	if bootstrappedCondition == nil || bootstrappedCondition.Status != gardencorev1beta1.ConditionTrue {
+		return
+	}
+	c.addSeedToQueue(obj)
+}
+
+func (c *Controller) addSeedToQueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.seedQueue.Add(key)
+}
+
+func (c *Controller) seedUpdate(oldObj, newObj interface{}) {
+	var (
+		newSeedObj = newObj.(*gardencorev1beta1.Seed)
+		oldSeedObj = oldObj.(*gardencorev1beta1.Seed)
+	)
+
+	newSeedBootstrappedCondition := helper.GetCondition(newSeedObj.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+	oldSeedBootstrappedCondition := helper.GetCondition(oldSeedObj.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+
+	if newSeedBootstrappedCondition != nil {
+		if oldSeedBootstrappedCondition == nil || seedChangedToBootstrapped(oldSeedBootstrappedCondition, newSeedBootstrappedCondition) {
```

Comment:
if newSeedBootstrappedCondition != nil and oldSeedBootstrappedCondition == nil the seed will be added to the queue without checking whether the newSeedBootstrappedCondition is True or not.

My suggestion is to use:
```
if seedChangedtoBootstrapped(newSeedBootstrappedCondition, oldSeedBootstrappedCondition) {
...
}
```
And then have the following definition for seedChangedToBootstrapped:
```
func seedChangedToBootstrapped(oldCondition, newCondition *gardencorev1beta1.Condition) bool {
	return newCondition != nil && newCondition.Status == gardencorev1beta1.ConditionTrue && (oldCondition== nil || oldCondition.Status != newCondition.Status)
}
```